### PR TITLE
Esil dfg lemon const folder

### PR DESCRIPTION
**Pull request by condret**
**Checklist**

- [ ] Closing issues: #issue
- [x] Mark this if you consider it ready to merge
- [x] I've added tests (optional)
- [ ] I wrote some lines in the [radare2book](https://github.com/radareorg/radare2book)

**Description**
The esil dfg filter already allows filtering for subexpressions, which would compute a specified register. This PR adds const folding to that. The following example shows filtering for eax, w/o const folding:
`4,!,3,ebx,:=,!,1,+,eax,:=  ===> 0x4,!,!,0x1,+,eax,:=`
Now the same with const folding:
`4,!,3,ebx,:=,!,1,+,eax,:=  ===> 0x2,eax,:=`

Const folding is done on the dfg graph, it is a function to tune up the dfg.

### CODE REVIEW

**Comment by condret** [comment]
https://twitter.com/r2gif/status/1292801285071294464/photo/1

**Review by trufae — CHANGES_REQUESTED** [review]
Code lgtm, but i would like to have a command to test it from r2 instead of C.

**Comment by condret** [comment]
Ok, I will add a command, I guess this could be useful at some point for r2pipe related things

File: libr/core/cmd_anal.c [hunk 1]
```diff
@@ -6072,6 +6073,98 @@ static void __anal_esil_function(RCore *core, ut64 addr) {
 	r_anal_esil_free (core->anal->esil);
 }
 
+static void print_esil_dfg_as_commands(RCore *core, RAnalEsilDFG *dfg) {
+	RListIter *iter, *ator;
+	RGraphNode *node, *edon;
+	RStrBuf *buf = r_strbuf_new ("");
```
[the remark this answers is not in the document]
Comment (condret) [reply libr/core/cmd_anal.c:6079]: yeah, I just copy pasted that from r_core_anal_esil_graph, but sure will do

[the remark this answers is not in the document]
Comment (trufae) [reply libr/core/cmd_anal.c]: Can you fix this and the other comments? Feel free to merge when green after this

File: libr/core/cmd_anal.c [hunk 2]
```diff
@@ -6072,6 +6073,105 @@ static void __anal_esil_function(RCore *core, ut64 addr) {
...
+	if (argv[0][0] == 'i' || argv[0][0] == 'v') {	//aegi & aegv
+		char *oprompt = strdup (r_config_get (core->config, "cmd.gprompt"));
+		r_config_set (core->config, "cmd.gprompt", "pi 1");
+		r_core_cmd0 (core, ".aeg*;aggv");
+		r_config_set (core->config, "cmd.gprompt", oprompt);
+		free (oprompt);
+		return;
+	}
+	if (argv[0][0] == '?') {
+		r_cons_printf ("Usage: aeg[cfiv*]\n");
```
Comment (trufae) [inline libr/core/cmd_anal.c:6145]: Use `r_core_cmd_help`

File: libr/core/cmd_anal.c [hunk 3]
```diff
@@ -6072,6 +6073,105 @@ static void __anal_esil_function(RCore *core, ut64 addr) {
...
+		if (R_STR_ISNOTEMPTY (esilstr)) {
+			RAnalEsilDFG *dfg = r_anal_esil_dfg_expr (core->anal, NULL, esilstr);
+			r_return_if_fail (dfg);
+			print_esil_dfg_as_commands (core, dfg);
+			r_anal_esil_dfg_free (dfg);
+		}
+		return;
+	}
+	if (argv[0][0] == 'i' || argv[0][0] == 'v') {	//aegi & aegv
+		char *oprompt = strdup (r_config_get (core->config, "cmd.gprompt"));
```
Comment (trufae) [inline libr/core/cmd_anal.c:6137]: There's RConfigHold for this.

```
 RConfigHold *hc = r_config_hold_new (core->config);
 r_config_hold_s (hc, "cmd.gprompt",  NULL);
....
r_config_hold_free (hc); 
```

File: libr/core/cmd_anal.c [hunk 4]
```diff
@@ -6072,6 +6073,105 @@ static void __anal_esil_function(RCore *core, ut64 addr) {
...
+	}
+	if (argc == 1) {	//aeg
+		RAnalEsilDFG *dfg = r_anal_esil_dfg_expr (core->anal, NULL, argv[0]);
+		r_return_if_fail (dfg);
+		print_esil_dfg_as_commands (core, dfg);
+		r_anal_esil_dfg_free (dfg);
+		return;
+	}
+	if (argv[0][0] == '*') {	//aeg*
+		//does this memleak?
```
Comment (trufae) [inline libr/core/cmd_anal.c:6124]: yes

File: libr/core/cmd_anal.c [hunk 5]
```diff
@@ -6072,6 +6073,105 @@ static void __anal_esil_function(RCore *core, ut64 addr) {
...
+		RAnalOp *aop = r_core_anal_op (core, core->offset, R_ANAL_OP_MASK_ESIL);
+		r_return_if_fail (aop);
+		const char *esilstr = r_strbuf_get (&aop->esil);
+		if (R_STR_ISNOTEMPTY (esilstr)) {
+			RAnalEsilDFG *dfg = r_anal_esil_dfg_expr (core->anal, NULL, esilstr);
+			r_return_if_fail (dfg);
```
Comment (trufae) [inline libr/core/cmd_anal.c:6130]: uhm is this possible to happen or to catch possible bugs in the code?

File: libr/core/cmd_anal.c [hunk 6]
```diff
@@ -6072,6 +6073,105 @@ static void __anal_esil_function(RCore *core, ut64 addr) {
...
+	if (argv[0][0] == '?') {
+		r_cons_printf ("Usage: aeg[cfiv*]\n");
+		r_cons_printf (" aeg  analyze current instruction as an esil graph\n");
+		r_cons_printf (" aegf analyze given expression and filter for register\n");
+		r_cons_printf (" aeg* analyze current instruction as an esil graph\n");
+		r_cons_printf (" aegv and launch the visual interactive mode (.aeg*;aggv == aegv)\n");
+		return;
+	}
+
+	RAnalEsilDFG *dfg = r_anal_esil_dfg_expr (core->anal, NULL, argv[1]);
```
Comment (trufae) [inline libr/core/cmd_anal.c:6153]: check if dfg is null

File: libr/core/cmd_anal.c [hunk 7]
```diff
@@ -6072,6 +6073,105 @@ static void __anal_esil_function(RCore *core, ut64 addr) {
...
+	if (argv[0][0] == 'f') {	//aegf
```
Comment (trufae) [inline libr/core/cmd_anal.c:6154]: use switch (argv[0][0]) {
case 'f': ...

File: libr/core/cmd_anal.c [hunk 8]
```diff
@@ -6072,6 +6073,105 @@ static void __anal_esil_function(RCore *core, ut64 addr) {
...
+		}
+	} else if (argv[0][0] == 'c') {
+		r_anal_esil_dfg_fold_const (core->anal, dfg);
+		if (argv[0][1] == 'f') {	//aegcf
+			RStrBuf *filtered = r_anal_esil_dfg_filter (dfg, argv[2]);
+			if (filtered) {
+				r_cons_printf ("%s\n", r_strbuf_get (filtered));
+				r_strbuf_free (filtered);
+			}
+		} else {	//aegc
```
Comment (trufae) [inline libr/core/cmd_anal.c:6168]: comments for grepping commands are `// "aegc"`

File: libr/core/cmd_anal.c [hunk 9]
```diff
@@ -6072,6 +6073,105 @@ static void __anal_esil_function(RCore *core, ut64 addr) {
...
+	}
+
+	RAnalEsilDFG *dfg = r_anal_esil_dfg_expr (core->anal, NULL, argv[1]);
+	if (argv[0][0] == 'f') {	//aegf
+		RStrBuf *filtered = r_anal_esil_dfg_filter (dfg, argv[2]);
+		if (filtered) {
+			r_cons_printf ("%s\n", r_strbuf_get (filtered));
+			r_strbuf_free (filtered);
+		}
+	} else if (argv[0][0] == 'c') {
```
Comment (trufae) [inline libr/core/cmd_anal.c:6160]: missing `// "aegc"`

File: libr/core/cmd_anal.c (hunk 7 above)
Comment (trufae) [inline libr/core/cmd_anal.c:6154]: fix syntax of comment `// "aegf"`

File: libr/util/rbtree.c [hunk 10]
```diff
@@ -488,6 +488,23 @@ R_API void *r_rbtree_cont_find(RContRBTree *tree, void *data, RContRBCmp cmp, vo
...
 
+// not a direct pendant to r_rbtree_first, but similar
+// returns first element in the tree, not an iter or a node
+R_API void *r_rbtree_cont_first(RContRBTree *tree) {
+	r_return_val_if_fail (tree, NULL);
+	if (!tree->root) {
+		// empty tree
+		return NULL;
+	}
+	RBIter iter = r_rbtree_first(&tree->root->node);
```
Comment (trufae) [inline libr/util/rbtree.c:499]: missing space before (

**Review by trufae — CHANGES_REQUESTED** [review]

File: libr/core/cmd_anal.c (hunk 2 above)
Comment (trufae) [reply libr/core/cmd_anal.c:6145]: also use switch(argv[0][0]) instead of many ifs

File: libr/core/cmd_anal.c (hunk 5 above)
Comment (condret) [reply libr/core/cmd_anal.c:6130]: it should only return NULL, if R_NEW0 fails

File: libr/core/cmd_anal.c (hunk 4 above)
Comment (condret) [reply libr/core/cmd_anal.c]: fixed

File: libr/core/cmd_anal.c (hunk 3 above)
Comment (condret) [reply libr/core/cmd_anal.c]: done

File: libr/core/cmd_anal.c (hunk 2 above)
Comment (condret) [reply libr/core/cmd_anal.c]: done

File: libr/core/cmd_anal.c [hunk 11]
```diff
@@ -6072,6 +6073,98 @@ static void __anal_esil_function(RCore *core, ut64 addr) {
 	r_anal_esil_free (core->anal->esil);
 }
 
+static void print_esil_dfg_as_commands(RCore *core, RAnalEsilDFG *dfg) {
+	RListIter *iter, *ator;
+	RGraphNode *node, *edon;
+	RStrBuf *buf = r_strbuf_new ("");
+	r_cons_printf ("ag-\n");
```
[the remark this answers is not in the document]
Comment (condret) [reply libr/core/cmd_anal.c]: done

File: libr/core/cmd_anal.c [hunk 12]
```diff
@@ -6072,6 +6073,98 @@ static void __anal_esil_function(RCore *core, ut64 addr) {
...
+	r_list_foreach (r_graph_get_nodes (dfg->flow), iter, node) {
+		const RAnalEsilDFGNode *enode = (RAnalEsilDFGNode *)node->data;
+		char *esc_str = r_str_escape (r_strbuf_get (enode->content));
+		r_strbuf_set (buf, esc_str);
+		if (enode->type == R_ANAL_ESIL_DFG_BLOCK_GENERATIVE) {
+			r_strbuf_prepend (buf, "generative:");
+		}
+		char *b64_buf = r_base64_encode_dyn (r_strbuf_get (buf), buf->len);
+		r_cons_printf ("agn %d base64:%s\n", enode->idx, b64_buf);
```
[the remark this answers is not in the document]
Comment (condret) [reply libr/core/cmd_anal.c:6111]: check added

File: libr/core/cmd_anal.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (condret) [reply libr/core/cmd_anal.c]: done

File: libr/core/cmd_anal.c (hunk 6 above)
Comment (condret) [reply libr/core/cmd_anal.c]: done

File: libr/core/cmd_anal.c (hunk 7 above)
Comment (condret) [reply libr/core/cmd_anal.c]: done

File: libr/core/cmd_anal.c (hunk 9 above)
Comment (condret) [reply libr/core/cmd_anal.c]: added

File: libr/util/rbtree.c (hunk 10 above)
Comment (condret) [reply libr/util/rbtree.c]: done

**Comment by trufae** [comment]
Party!

**Pull request merged**